### DISCUSSION TITLE
How can I send data between pages in NextJS?

### DISCUSSION THREAD

**Discussion by OwaisSiddiqui** [vercel/next.js#13319] [Help] [answered]
Hi. I am building a website using NextJS (React on the server-side).

I started by using npx create-next-app but then I also added a custom ExpressJS server. I have a page (in the pages directory) where a user inputs a city name. Then, a Javascript file redirects the user (using window.location.href = "<page_name>") to the next page which requires the user data of the city name.

How would I be able to receive the data, the city name, on the next page (<page_name>) specifically using NextJS (and not something like query parameters in the URL, but please tell me if this is the better solution)?

I was thinking of using the NextJS API, however, when I make a fetch post request from the JavaScript file (linked to index.js) and receive it in a file in the API folder, I am not sure how to store the data from the request so that when another page gives a request to the API file, I can send the data the user gave.

Again, I don't know if using the AP I is the best or even valid choice. I am still new to NextJS and React. Here is my directory structure:
`project_name
  .next
  components
      // components
  node_modules
      // node modules
  pages
      api
          // Stuff
      // pages
  public
      // stuff
  // .gitignore, server.js, package-json etc.`

**Comment by leerob** [comment, marked as answer]
1. Instead of redirecting with `window.location.href`, consider using Next's [router](https://nextjs.org/docs/routing/introduction).
2. To pass data between these pages, you could use [query parameters](https://branch.io/glossary/query-parameters/).
3. If you need persistent storage but don't want to use a database, try this [`useLocalStorage` hook](https://usehooks.com/useLocalStorage/).

**Reply by OwaisSiddiqui** [reply]
Okay, thanks. About the first point, I have the logic of handling the city name input from a Google Autocomplete Search bar, and I have functions for that inside a JS file. How would I be able to use Next's router within the JS file?

**Reply by leerob** [reply]
As far as I know, you wouldn't be able to use Next's router outside of the context of React (e.g., vanilla JS file for something third-party). Can you make the search happen `onChange` for the autocomplete input?

**Reply by OwaisSiddiqui** [reply]
I am getting an invalid hook error, here is some sort of pseudo-code of the problem:
Error message:
```
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
```
index.js:
```
import { useEffect, Component } from 'react'
import { useRouter } from 'next/router'
// Some other imports

class Home extends Component {
    handleSearch(event) {
         const router = useRouter()
         // Google autocomplete JS logic
         if (place clicked) {
             router.push({
                 pathname: '/app',
                query: { place: 'City' },
             }) 
         }
    }

    render() {
         return (
             <input /> // Autocomplete input        
         )
    }
}

export default Home
```

**Reply by leerob** [reply]
I'd recommend reading through these React docs first. 

- [Function and Class Components](https://reactjs.org/docs/components-and-props.html#function-and-class-components)
- [Hooks and Function Components](https://reactjs.org/docs/hooks-state.html#hooks-and-function-components)

The issue is that you're trying to use `useRouter` inside of a class component. You'd need to convert that file to a functional component. 

```js
import { useRouter } from 'next/router';

export const Home = () => {
  const router = useRouter();
  const handleSearch = (event) => {
    if (clicked) {
      router.push({
        pathname: '/app',
        query: { place: 'City' }
      });
    }
  };

  return (
    <input /> // Autocomplete input
  );
};
```

**Reply by OwaisSiddiqui** [reply]
Perfect, it works! I am new to both Next and React JS. I appreciate the links you provided, I will definitely get to read them when I get the time. Thanks for the help!

**Comment by mohanas-scg** [comment]
Hi, @leerob how to pass large data without showing that in the URL?

**Reply by phoneticallySAARTHaK** [reply]
Either I'm not looking hard enough, or this feature doesn't exist

**Reply by leerob** [reply]
This is now much easier, and built in natively, with the App Router

https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating

**Reply by phoneticallySAARTHaK** [reply]
@leerob  that's pretty cool. Is it possible to create pages(routes) for specific HTTP Methods?
Like how old-school HTML Form routing works, rendering stuff via a POST call.

**Reply by leerob** [reply]
You can create [Router Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) to make an endpoint that accepts specific HTTP verbs. But with [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations), you can actually skip the manual creation of the API and instead just call a function. It makes the API for you. Neat stuff!

**Reply by krishnagptamcs** [reply]
> This is now much easier, and built in natively, with the App Router
> 
> https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating

but how ???? @leerob

**Reply by krishnagptamcs** [reply]
i did'nt found in the docs you share , if possible pls share with an example @leerob
